src/group_theory/nilpotent.lean

```diff
@@ -574,6 +576,101 @@ instance nilpotent_quotient_of_nilpotent (H : subgroup G) [H.normal] [h : is_nil
+  comap (mk' H₁) (center (G ⧸ H₁)) = comap (mk' H₂) (center (G ⧸ H₂)) :=
+  by { unfreezingI { subst h, } }
+
+lemma comap_upper_central_series_quotient_center (n : ℕ) :
+  comap (mk' (center G)) (upper_central_series (G ⧸ center G) n) = upper_central_series G n.succ :=
+begin
+  induction n with n ih,
+  { simp, },
+  { let Hn := upper_central_series (G ⧸ center G) n,
+    calc comap (mk' (center G)) (upper_central_series (G ⧸ center G) n.succ)
+        = comap (mk' (center G)) (upper_central_series_step Hn)
+        : rfl
+    ... = comap (mk' (center G)) (comap (mk' Hn) (center ((G ⧸ center G) ⧸ Hn)))
+        : by rw upper_central_series_step_eq_comap_center
+    ... = comap (mk' (comap (mk' (center G)) Hn)) (center (G ⧸ (comap (mk' (center G)) Hn)))
+        : comap_comap_center
+    ... = comap (mk' (upper_central_series G n.succ)) (center (G ⧸ upper_central_series G n.succ))
+        : comap_center_subst ih
+    ... = upper_central_series_step (upper_central_series G n.succ)
+        : symm (upper_central_series_step_eq_comap_center _)
+    ... = upper_central_series G n.succ.succ
+        : rfl, },
+end
+
+lemma nilpotency_class_zero_iff_subsingleton [is_nilpotent G] :
+  group.nilpotency_class G = 0 ↔ subsingleton G :=
+by simp [group.nilpotency_class, nat.find_eq_zero, subsingleton_iff_bot_eq_top]
+
+section classical
+
+open_locale classical
+
+/-- Quotienting the `center G` reduces the nilpotency class by 1 -/
+lemma nilpotency_class_quotient_center [hH : is_nilpotent G] :
+  group.nilpotency_class (G ⧸ center G) = group.nilpotency_class G - 1 :=
+begin
+  generalize hn : group.nilpotency_class G = n,
+  rcases n with rfl | n,
+  { simp [nilpotency_class_zero_iff_subsingleton] at *,
+    haveI := hn,
+    apply_instance, },
+  { suffices : group.nilpotency_class (G ⧸ center G) = n, by simpa,
+    apply le_antisymm,
+    { apply upper_central_series_eq_top_iff_nilpotency_class_le.mp,
+      apply (@comap_injective G _ _ _ (mk' (center G)) (surjective_quot_mk _)),
+      rw [ comap_upper_central_series_quotient_center, comap_top, ← hn],
+      exact upper_central_series_nilpotency_class, },
+    { apply le_of_add_le_add_right,
+      calc n + 1 = n.succ : rfl
+        ... = group.nilpotency_class G : symm hn
+        ... ≤ group.nilpotency_class (G ⧸ center G) + 1
+            : nilpotency_class_le_of_ker_le_center _ (le_of_eq (ker_mk _)) _, } }
+end
+
+end classical
+
+/-- A custom induction principle for nilpotent groups. The base case is a trivial group
+(`subsingleton G`), and in the inductoin step, one can assume the hypothesis for
+the group quotiented by its center.
+-/
+@[elab_as_eliminator]
+lemma nilpotent_center_quotient_ind
+  {P : Π G [group G], by exactI ∀ [is_nilpotent G], Prop}
+  (G : Type*)
+  [group G]
+  [is_nilpotent G]
+  (hbase : ∀ G [group G], by exactI ∀ [is_nilpotent G], by exactI ∀ [subsingleton G], P G)
+  (hstep : ∀ G [group G], by exactI ∀ [is_nilpotent G], by exactI ∀ (ih : P (G ⧸ center G)), P G) :
```

Comment:
A matter of taste, feel free to ignore.
```suggestion
  (hstep : ∀ G [group G], by exactI ∀ [is_nilpotent G], by exactI P (G ⧸ center G) → P G) :
```

Reply:
I was assuming that using a name here will give a hint to `induction` as to how to name the assumption. I’ll check if that’s actually the case.

Reply:
But we have the policy that mathlib shouldn't use auto-generated names anyway. So that doesn't matter too much, I think.

Reply:
As I thought: this is how `induction` gets the names from, so I’ll leave it like this.